zaza/openstack/charm_tests/series_upgrade/tests.py

```diff
@@ -215,5 +217,166 @@ def setUpClass(cls):
+            for application, app_details in applications.items():
+                # Defaults
+                origin = "openstack-origin"
+                pause_non_leader_subordinate = True
+                pause_non_leader_primary = True
+                post_upgrade_functions = []
+                name = upgrade_utils.extract_charm_name_from_url(
+                    app_details['charm'])
+                if name not in group and application not in group:
+                    if group_name != "deferred" and \
+                            name not in upgrade_groups['deferred']:
+                        upgrade_groups['deferred'].append(name)
+                    continue
+                if group_name != "deferred" and \
+                        name in upgrade_groups['deferred']:
+                    upgrade_groups['deferred'].remove(name)
+                # Skip subordinates
+                if app_details["subordinate-to"]:
+                    continue
+                if "easyrsa" in app_details["charm"]:
+                    logging.warn(
+                        "Skipping series upgrade of easyrsa Bug #1850121")
+                    continue
+                if "etcd" in app_details["charm"]:
+                    logging.warn(
+                        "Skipping series upgrade of easyrsa Bug #1850124")
+                    continue
+                logging.warn("About to upgrade {}".format(application))
+                if "percona-cluster" in app_details["charm"]:
+                    origin = "source"
+                    pause_non_leader_primary = True
+                    pause_non_leader_subordinate = True
+                if "rabbitmq-server" in app_details["charm"]:
+                    origin = "source"
+                    pause_non_leader_primary = True
+                    pause_non_leader_subordinate = False
+                if "nova-compute" in app_details["charm"]:
+                    pause_non_leader_primary = False
+                    pause_non_leader_subordinate = False
+                if "ceph" in app_details["charm"]:
+                    origin = "source"
+                    pause_non_leader_primary = False
+                    pause_non_leader_subordinate = False
+                if "designate-bind" in app_details["charm"]:
+                    origin = None
+                if "tempest" in app_details["charm"]:
+                    origin = None
+                if "memcached" in app_details["charm"]:
+                    origin = None
+                    pause_non_leader_primary = False
+                    pause_non_leader_subordinate = False
+                if "vault" in app_details["charm"]:
+                    origin = None
+                    pause_non_leader_primary = False
+                    pause_non_leader_subordinate = True
+                    post_upgrade_functions = [
+                        ('zaza.openstack.charm_tests.vault.setup.'
+                         'mojo_unseal_by_unit')]
```

Comment:
I know this exists in the serial upgrade code but I'd really like to tidy up this monolithic 'if' statement. I was thinking about something like:

```
import collections

import copy

def get_charm_settings():
    default = {
        'origin': 'openstack-origin',
        'pause_non_leader_subordinate': True,
        'pause_non_leader_primary': True,
        'post_upgrade_functions': []}

    _app_settings = collections.defaultdict(lambda: default)
    exceptions = {
        'rabbitmq-server': {
            'origin': 'source',
            'pause_non_leader_subordinate': False},
        'percona-cluster': {'origin': 'source'},
        'nova-compute' : {
            'pause_non_leader_primary': False,
            'pause_non_leader_subordinate': False}}
    for key, value in exceptions.items():
        _app_settings[key] = copy.deepcopy(default)
        _app_settings[key].update(value)
    return _app_settings

settings = get_charm_settings()
for key in settings.keys():
    if key in app_details["charm"]:
        app_settings = settings[key]
        break
else:
    app_settings = settings[app_details["charm"]]
...
upgrade_group.append(
   series_upgrade_utils.async_series_upgrade_application(
￼      application,
￼      pause_non_leader_primary=app_settings['pause_non_leader_primary'],
￼      pause_non_leader_subordinate=pause,
￼      from_series=self.from_series,
￼      to_series=self.to_series,
￼      origin=app_settings['origin'],
￼      completed_machines=completed_machines,
￼      workaround_script=self.workaround_script,
￼      files=self.files,
￼      post_upgrade_functions=app_settings['post_upgrade_functions']))

```

what do you think?